awx_collection/plugins/modules/tower_user.py

```diff
@@ -106,59 +107,50 @@
     tower_config_file: "~/tower_cli.cfg"
 '''
 
-from ..module_utils.ansible_tower import TowerModule, tower_auth_config, tower_check_mode
-
-try:
-    import tower_cli
-    import tower_cli.exceptions as exc
-
-    from tower_cli.conf import settings
-except ImportError:
-    pass
+from ..module_utils.tower_api import TowerModule
 
 
 def main():
+    # Any additional arguments that are not fields of the item can be added here
     argument_spec = dict(
         username=dict(required=True),
         first_name=dict(),
         last_name=dict(),
         password=dict(no_log=True),
-        email=dict(required=True),
+        email=dict(required=False),
         superuser=dict(type='bool', default=False),
         auditor=dict(type='bool', default=False),
         state=dict(choices=['present', 'absent'], default='present'),
     )
 
-    module = TowerModule(argument_spec=argument_spec, supports_check_mode=True)
+    # Create a module for ourselves
+    module = TowerModule(argument_spec=argument_spec, supports_check_mode=True, required_if=[['state', 'present', ['email']]])
 
-    username = module.params.get('username')
-    first_name = module.params.get('first_name')
-    last_name = module.params.get('last_name')
-    password = module.params.get('password')
-    email = module.params.get('email')
-    superuser = module.params.get('superuser')
-    auditor = module.params.get('auditor')
+    # Extract our parameters
     state = module.params.get('state')
-
-    json_output = {'username': username, 'state': state}
-
-    tower_auth = tower_auth_config(module)
-    with settings.runtime_values(**tower_auth):
-        tower_check_mode(module)
-        user = tower_cli.get_resource('user')
-        try:
-            if state == 'present':
-                result = user.modify(username=username, first_name=first_name, last_name=last_name,
-                                     email=email, password=password, is_superuser=superuser,
-                                     is_system_auditor=auditor, create_on_missing=True)
-                json_output['id'] = result['id']
-            elif state == 'absent':
-                result = user.delete(username=username)
-        except (exc.ConnectionError, exc.BadRequest, exc.AuthError) as excinfo:
-            module.fail_json(msg='Failed to update the user: {0}'.format(excinfo), changed=False)
-
-    json_output['changed'] = result['changed']
-    module.exit_json(**json_output)
+    user_fields = {
+        'username': module.params.get('username'),
+        'first_name': module.params.get('first_name'),
+        'last_name': module.params.get('last_name'),
+        'password': module.params.get('password'),
```

Comment:
how are null values handled?? Does the `create_or_update_if_needed` method ignore them? I'm interested in the case where the module used, and non-required parameters are not provided.

Reply:
In the `update_if_needed` method in the new module utils, the code below should be taking care of this:

```
            for field in new_item:
                existing_field = existing_item.get(field, None)
                new_field = new_item.get(field, None)
                # If the two items don't match and we are not comparing '' to None
                if existing_field != new_field and not (existing_field in (None, '') and new_field == ''):
                    # Something doesn't match so let's update it
                    needs_update = True
                    break
```

At the module level, we should only be adding in the parameters if the optional ones are set (like what you did most recently in `inventory_source`).